File: src/main/java/org/cimsbioko/data/LocationGateway.java
package org.cimsbioko.data;

import android.content.ContentValues;
import android.database.Cursor;

import org.cimsbioko.App;
import org.cimsbioko.R;
import org.cimsbioko.model.core.Location;
import org.jetbrains.annotations.NotNull;

import static org.cimsbioko.App.Locations.COLUMN_LOCATION_ATTRS;
import static org.cimsbioko.App.Locations.COLUMN_LOCATION_DESCRIPTION;
import static org.cimsbioko.App.Locations.COLUMN_LOCATION_EXTID;
import static org.cimsbioko.App.Locations.COLUMN_LOCATION_HIERARCHY_UUID;
import static org.cimsbioko.App.Locations.COLUMN_LOCATION_LATITUDE;
import static org.cimsbioko.App.Locations.COLUMN_LOCATION_LONGITUDE;
import static org.cimsbioko.App.Locations.COLUMN_LOCATION_NAME;
import static org.cimsbioko.App.Locations.COLUMN_LOCATION_UUID;
import static org.cimsbioko.navconfig.Hierarchy.HOUSEHOLD;
import static org.cimsbioko.data.CursorConvert.extractString;

/**
 * Convert Locations to and from database.  Location-specific queries.
 */
public class LocationGateway extends Gateway<Location> {

    private static final LocationEntityConverter ENTITY_CONVERTER = new LocationEntityConverter();
    private static final LocationWrapperConverter WRAPPER_CONVERTER = new LocationWrapperConverter();
    private static final LocationContentValuesConverter CONTENT_VALUES_CONVERTER = new LocationContentValuesConverter();

    LocationGateway() {
        super(App.Locations.CONTENT_ID_URI_BASE, COLUMN_LOCATION_UUID);
    }

    public Query<Location> findByHierarchy(String hierarchyId) {
        return new Query<>(this, tableUri, COLUMN_LOCATION_HIERARCHY_UUID, hierarchyId, COLUMN_LOCATION_EXTID);
    }

    @Override
    String getId(Location entity) {
        return entity.getUuid();
    }

    @Override
    public CursorConverter<Location> getEntityConverter() {
        return ENTITY_CONVERTER;
    }

    @Override
    public CursorConverter<DataWrapper> getWrapperConverter() {
        return WRAPPER_CONVERTER;
    }

    @Override
    ContentValuesConverter<Location> getContentValuesConverter() {
        return CONTENT_VALUES_CONVERTER;
    }
}

class LocationEntityConverter implements CursorConverter<Location> {
    @Override
    @NotNull
    public Location convert(@NotNull Cursor c) {
        Location location = new Location();
        location.setUuid(extractString(c, COLUMN_LOCATION_UUID));
        location.setExtId(extractString(c, COLUMN_LOCATION_EXTID));
        location.setHierarchyUuid(extractString(c, COLUMN_LOCATION_HIERARCHY_UUID));
        location.setLatitude(extractString(c, COLUMN_LOCATION_LATITUDE));
        location.setLongitude(extractString(c, COLUMN_LOCATION_LONGITUDE));
        location.setName(extractString(c, COLUMN_LOCATION_NAME));
        location.setDescription(extractString(c, COLUMN_LOCATION_DESCRIPTION));
        location.setLongitude(extractString(c, COLUMN_LOCATION_LONGITUDE));
        location.setLatitude(extractString(c, COLUMN_LOCATION_LATITUDE));
        location.setAttrs(extractString(c, COLUMN_LOCATION_ATTRS));
        return location;
    }
}

class LocationWrapperConverter implements CursorConverter<DataWrapper> {
    @Override
    @NotNull
    public DataWrapper convert(@NotNull Cursor c) {
        DataWrapper dataWrapper = new DataWrapper(
                extractString(c, COLUMN_LOCATION_UUID),
                HOUSEHOLD,
                extractString(c, COLUMN_LOCATION_EXTID),
                extractString(c, COLUMN_LOCATION_NAME)
        );
        dataWrapper.getStringsPayload().put(R.string.location_description_label, extractString(c, COLUMN_LOCATION_DESCRIPTION));
        return dataWrapper;
    }
}

class LocationContentValuesConverter implements ContentValuesConverter<Location> {

    @Override
    @NotNull
    public ContentValues toContentValues(@NotNull Location location) {
        ContentValues contentValues = new ContentValues();
        contentValues.put(COLUMN_LOCATION_UUID, location.getUuid());
        contentValues.put(COLUMN_LOCATION_EXTID, location.getExtId());
        contentValues.put(COLUMN_LOCATION_HIERARCHY_UUID, location.getHierarchyUuid());
        contentValues.put(COLUMN_LOCATION_LATITUDE, location.getLatitude());
        contentValues.put(COLUMN_LOCATION_LONGITUDE, location.getLongitude());
        contentValues.put(COLUMN_LOCATION_NAME, location.getName());
        contentValues.put(COLUMN_LOCATION_DESCRIPTION, location.getDescription());
        contentValues.put(COLUMN_LOCATION_LONGITUDE, location.getLongitude());
        contentValues.put(COLUMN_LOCATION_LATITUDE, location.getLatitude());
        contentValues.put(COLUMN_LOCATION_ATTRS, location.getAttrs());
        return contentValues;
    }
}

File: src/main/kotlin/org/cimsbioko/data/LocationGateway.kt
package org.cimsbioko.data

import android.content.ContentValues
import android.database.Cursor
import org.cimsbioko.App.Locations
import org.cimsbioko.R
import org.cimsbioko.data.CursorConvert.extractString
import org.cimsbioko.model.core.Location
import org.cimsbioko.navconfig.Hierarchy

/**
 * Convert Locations to and from database.  Location-specific queries.
 */
class LocationGateway internal constructor()
    : Gateway<Location>(Locations.CONTENT_ID_URI_BASE, Locations.COLUMN_LOCATION_UUID) {

    fun findByHierarchy(hierarchyId: String): Query<Location> =
            Query(this, tableUri, Locations.COLUMN_LOCATION_HIERARCHY_UUID, hierarchyId, Locations.COLUMN_LOCATION_EXTID)

    override fun getId(entity: Location): String = entity.uuid!!

    override val entityConverter: CursorConverter<Location> by lazy { LocationEntityConverter() }
    override val wrapperConverter: CursorConverter<DataWrapper> by lazy { LocationWrapperConverter() }
    override val contentValuesConverter: ContentValuesConverter<Location> by lazy { LocationContentValuesConverter() }
}

private class LocationEntityConverter : CursorConverter<Location> {
    override fun convert(c: Cursor): Location = Location().apply {
        uuid = extractString(c, Locations.COLUMN_LOCATION_UUID)
        extId = extractString(c, Locations.COLUMN_LOCATION_EXTID)
        hierarchyUuid = extractString(c, Locations.COLUMN_LOCATION_HIERARCHY_UUID)
        latitude = extractString(c, Locations.COLUMN_LOCATION_LATITUDE)
        longitude = extractString(c, Locations.COLUMN_LOCATION_LONGITUDE)
        name = extractString(c, Locations.COLUMN_LOCATION_NAME)
        description = extractString(c, Locations.COLUMN_LOCATION_DESCRIPTION)
        longitude = extractString(c, Locations.COLUMN_LOCATION_LONGITUDE)
        latitude = extractString(c, Locations.COLUMN_LOCATION_LATITUDE)
        attrs = extractString(c, Locations.COLUMN_LOCATION_ATTRS)
    }
}

private class LocationWrapperConverter : CursorConverter<DataWrapper> {
    override fun convert(c: Cursor): DataWrapper = DataWrapper(
            extractString(c, Locations.COLUMN_LOCATION_UUID)!!,
            Hierarchy.HOUSEHOLD,
            extractString(c, Locations.COLUMN_LOCATION_EXTID)!!,
            extractString(c, Locations.COLUMN_LOCATION_NAME)!!
    ).apply {
        stringsPayload[R.string.location_description_label] = extractString(c, Locations.COLUMN_LOCATION_DESCRIPTION)
    }
}

private class LocationContentValuesConverter : ContentValuesConverter<Location> {
    override fun toContentValues(entity: Location): ContentValues = ContentValues().apply {
        with(entity) {
            put(Locations.COLUMN_LOCATION_UUID, uuid)
            put(Locations.COLUMN_LOCATION_EXTID, extId)
            put(Locations.COLUMN_LOCATION_HIERARCHY_UUID, hierarchyUuid)
            put(Locations.COLUMN_LOCATION_LATITUDE, latitude)
            put(Locations.COLUMN_LOCATION_LONGITUDE, longitude)
            put(Locations.COLUMN_LOCATION_NAME, name)
            put(Locations.COLUMN_LOCATION_DESCRIPTION, description)
            put(Locations.COLUMN_LOCATION_LONGITUDE, longitude)
            put(Locations.COLUMN_LOCATION_LATITUDE, latitude)
            put(Locations.COLUMN_LOCATION_ATTRS, attrs)
        }
    }
}